File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
    var stack []byte

    for _, c := range s {
        if (c == '(' || c == '{' || c == '[') {
            stack = append(stack, byte(c))
        } else {
            if (len(stack) == 0) {
                return false
            } else {
                if (c == ')' && stack[len(stack)-1] == '(') {
                    stack = stack[:len(stack)-1]
                } else if (c == '}' && stack[len(stack)-1] == '{') {
                    stack = stack[:len(stack)-1]
                } else if (c == ']' && stack[len(stack)-1] == '[') {
                    stack = stack[:len(stack)-1]
                } else {
                    return false
                }
            }
        }
    }
    return len(stack) == 0
}

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> st = new ArrayDeque<>();
        int n = s.length();

        for (int i = 0; i < n; i++) {
            Character key = s.charAt(i);
            if (key == '(' || key == '{' || key == '[') {
                st.add(key);
                continue;
            }

            if (st.isEmpty()) {
                return false;
            }

            if (key == ')' && st.peekLast() == '(') {
                st.removeLast();
                continue;
            }

            if (key == '}' && st.peekLast() == '{') {
                st.removeLast();
                continue;
            }

            if (key == ']' && st.peekLast() == '[') {
                st.removeLast();
                continue;
            }
            return false;
        }
        return st.size() == 0;

    }
}